<a href="https://colab.research.google.com/github/Shivamkumar-eng/Codsoft_task1/blob/main/student_performance_predictor_for_eduquest_coaching.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [140]:
# import library
import pandas as pd


In [141]:
# Read dataset
df= pd.read_csv('/content/Student Performance Predictor for EduQuest Coaching.csv')

In [142]:
df.head()

gender  age parental_education  family_income internet_access  \
0    Male   14            Diploma       29591.84             Yes   
1    Male   16            Diploma       35222.46             Yes   
2    Male   14            Diploma       29730.93             Yes   
3  Female   15            Diploma       29074.09             Yes   
4    Male   19           Graduate       42305.33             Yes   

   previous_exam_score  attendance_rate  homework_completion_rate  \
0                93.25            91.01                     76.56   
1                85.06            83.85                     64.82   
2                79.93            86.71                     75.02   
3                98.94            88.04                     55.50   
4                64.54            61.22                     98.85   

   class_participation_score  number_of_absences extra_curricular_involvement  \
0                       9.52                   1                     Moderate   
1                       3.42                   2                     Moderate   
2                       3.75                   1                          Low   
3                       6.95                   0                     Moderate   
4                       6.03                   5                     Moderate   

   learning_hours_per_week tutor_support  final_exam_score  
0                    12.08            No            100.00  
1                    11.81           Yes             80.81  
2                    11.08           Yes             77.96  
3                     6.65           Yes             98.86  
4                     9.26            No             64.64

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   gender                        1000 non-null   object 
 1   age                           1000 non-null   int64  
 2   parental_education            1000 non-null   object 
 3   family_income                 1000 non-null   float64
 4   internet_access               1000 non-null   object 
 5   previous_exam_score           1000 non-null   float64
 6   attendance_rate               1000 non-null   float64
 7   homework_completion_rate      1000 non-null   float64
 8   class_participation_score     1000 non-null   float64
 9   number_of_absences            1000 non-null   int64  
 10  extra_curricular_involvement  628 non-null    object 
 11  learning_hours_per_week       1000 non-null   float64
 12  tutor_support                 1000 non-null   object 
 13  fina

# In df.info() only  extra_curricular_involvement  has  628 value
## which means is contain missing values so we will ignore this

In [144]:
df= df.dropna()

# throug df.dropna() we  drop the  misssing value

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 628 entries, 0 to 999
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   gender                        628 non-null    object 
 1   age                           628 non-null    int64  
 2   parental_education            628 non-null    object 
 3   family_income                 628 non-null    float64
 4   internet_access               628 non-null    object 
 5   previous_exam_score           628 non-null    float64
 6   attendance_rate               628 non-null    float64
 7   homework_completion_rate      628 non-null    float64
 8   class_participation_score     628 non-null    float64
 9   number_of_absences            628 non-null    int64  
 10  extra_curricular_involvement  628 non-null    object 
 11  learning_hours_per_week       628 non-null    float64
 12  tutor_support                 628 non-null    object 
 13  final_exam

# NOW WE DESCRIBE THE DATA THROUGH df.describe()
# which ignore the non numeric or object data

In [146]:
df.describe()

age  family_income  previous_exam_score  attendance_rate  \
count  628.000000     628.000000           628.000000       628.000000   
mean    16.498408   39458.801736            70.201863        80.757994   
std      1.667304    9776.354037            14.059668        11.483176   
min     14.000000   14302.140000            30.000000        60.050000   
25%     15.000000   33004.172500            60.900000        71.220000   
50%     17.000000   39698.730000            70.200000        81.220000   
75%     18.000000   46414.660000            79.760000        90.590000   
max     19.000000   63548.870000           100.000000        99.990000   

       homework_completion_rate  class_participation_score  \
count                628.000000                 628.000000   
mean                  74.802150                   4.817675   
std                   14.508351                   2.851535   
min                   50.050000                   0.030000   
25%                   61.950000                   2.172500   
50%                   75.595000                   4.905000   
75%                   87.267500                   7.182500   
max                   99.990000                   9.990000   

       number_of_absences  learning_hours_per_week  final_exam_score  
count          628.000000               628.000000        628.000000  
mean             2.006369                10.037500         76.093455  
std              1.449839                 3.126036         11.122763  
min              0.000000                 0.910000         39.840000  
25%              1.000000                 7.977500         68.140000  
50%              2.000000                10.015000         76.845000  
75%              3.000000                12.142500         84.042500  
max              8.000000                18.420000        100.000000

# THROUGH df.describe(include='all') we force to include all

In [147]:
df.describe(include='all')

gender         age parental_education  family_income internet_access  \
count      628  628.000000                628     628.000000             628   
unique       2         NaN                  4            NaN               2   
top     Female         NaN            Diploma            NaN             Yes   
freq       316         NaN                198            NaN             543   
mean       NaN   16.498408                NaN   39458.801736             NaN   
std        NaN    1.667304                NaN    9776.354037             NaN   
min        NaN   14.000000                NaN   14302.140000             NaN   
25%        NaN   15.000000                NaN   33004.172500             NaN   
50%        NaN   17.000000                NaN   39698.730000             NaN   
75%        NaN   18.000000                NaN   46414.660000             NaN   
max        NaN   19.000000                NaN   63548.870000             NaN   

        previous_exam_score  attendance_rate  homework_completion_rate  \
count            628.000000       628.000000                628.000000   
unique                  NaN              NaN                       NaN   
top                     NaN              NaN                       NaN   
freq                    NaN              NaN                       NaN   
mean              70.201863        80.757994                 74.802150   
std               14.059668        11.483176                 14.508351   
min               30.000000        60.050000                 50.050000   
25%               60.900000        71.220000                 61.950000   
50%               70.200000        81.220000                 75.595000   
75%               79.760000        90.590000                 87.267500   
max              100.000000        99.990000                 99.990000   

        class_participation_score  number_of_absences  \
count                  628.000000          628.000000   
unique                        NaN                 NaN   
top                           NaN                 NaN   
freq                          NaN                 NaN   
mean                     4.817675            2.006369   
std                      2.851535            1.449839   
min                      0.030000            0.000000   
25%                      2.172500            1.000000   
50%                      4.905000            2.000000   
75%                      7.182500            3.000000   
max                      9.990000            8.000000   

       extra_curricular_involvement  learning_hours_per_week tutor_support  \
count                           628               628.000000           628   
unique                            3                      NaN             2   
top                             Low                      NaN           Yes   
freq                            311                      NaN           348   
mean                            NaN                10.037500           NaN   
std                             NaN                 3.126036           NaN   
min                             NaN                 0.910000           NaN   
25%                             NaN                 7.977500           NaN   
50%                             NaN                10.015000           NaN   
75%                             NaN                12.142500           NaN   
max                             NaN                18.420000           NaN   

        final_exam_score  
count         628.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean           76.093455  
std            11.122763  
min            39.840000  
25%            68.140000  
50%            76.845000  
75%            84.042500  
max           100.000000

# //ENCODING

In [148]:
# Here this syntax give the catagories of data
df['gender'].value_counts()

gender
Female    316
Male      312
Name: count, dtype: int64

**NOW WE  GOING TO REPLACE THIS INTO 0 AND 1**



In [149]:
df['gender']= df['gender'].replace({'Male':0, 'Female':1})

In [150]:
df['gender'].value_counts()

gender
1    316
0    312
Name: count, dtype: int64

In [151]:
df['parental_education'].value_counts()

parental_education
Diploma         198
Graduate        182
High School     181
Postgraduate     67
Name: count, dtype: int64

**NOW WE CONVERT  parental_education strings INTO INTIGER FORM INTO A SERIAL MANER**

In [152]:
df['parental_education']=df['parental_education'].replace({'High School':0,'Diploma':1,'Graduate':2,'Postgraduate':3})

In [153]:
df['internet_access']=df['internet_access'].replace({'Yes':1,'No':0})

In [154]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 628 entries, 0 to 999
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   gender                        628 non-null    int64  
 1   age                           628 non-null    int64  
 2   parental_education            628 non-null    int64  
 3   family_income                 628 non-null    float64
 4   internet_access               628 non-null    int64  
 5   previous_exam_score           628 non-null    float64
 6   attendance_rate               628 non-null    float64
 7   homework_completion_rate      628 non-null    float64
 8   class_participation_score     628 non-null    float64
 9   number_of_absences            628 non-null    int64  
 10  extra_curricular_involvement  628 non-null    object 
 11  learning_hours_per_week       628 non-null    float64
 12  tutor_support                 628 non-null    object 
 13  final_exam

# extra_curricular_involvement and tutor_support are also  object so convert this in numeric for as others

In [155]:
df['extra_curricular_involvement'].value_counts()

extra_curricular_involvement
Low         311
Moderate    224
High         93
Name: count, dtype: int64

In [156]:
df['extra_curricular_involvement']= df['extra_curricular_involvement'].replace({'Low':0,'Moderate':1,'High':2})

In [157]:
df['tutor_support'].value_counts()

tutor_support
Yes    348
No     280
Name: count, dtype: int64

**now we going to check again**



In [158]:
df['tutor_support']=df['tutor_support'].replace({'Yes':1,'No':0})

In [159]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 628 entries, 0 to 999
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   gender                        628 non-null    int64  
 1   age                           628 non-null    int64  
 2   parental_education            628 non-null    int64  
 3   family_income                 628 non-null    float64
 4   internet_access               628 non-null    int64  
 5   previous_exam_score           628 non-null    float64
 6   attendance_rate               628 non-null    float64
 7   homework_completion_rate      628 non-null    float64
 8   class_participation_score     628 non-null    float64
 9   number_of_absences            628 non-null    int64  
 10  extra_curricular_involvement  628 non-null    int64  
 11  learning_hours_per_week       628 non-null    float64
 12  tutor_support                 628 non-null    int64  
 13  final_exam

***// DEFINE X AND Y ***

In [160]:
df.columns

Index(['gender', 'age', 'parental_education', 'family_income',
       'internet_access', 'previous_exam_score', 'attendance_rate',
       'homework_completion_rate', 'class_participation_score',
       'number_of_absences', 'extra_curricular_involvement',
       'learning_hours_per_week', 'tutor_support', 'final_exam_score'],
      dtype='object')

In [161]:
y=df['final_exam_score']
x=df[['gender', 'age', 'parental_education', 'family_income',
       'internet_access', 'previous_exam_score', 'attendance_rate',
       'homework_completion_rate', 'class_participation_score',
       'number_of_absences', 'extra_curricular_involvement',
       'learning_hours_per_week', 'tutor_support',]]

**NOW WE GOING TO SPILIT THE DATA INTO TRAIN AND TEST**

In [162]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,random_state=2529)

**//NOW WE GOING TO SELECT THE MODEL**

In [163]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

**//TRAIN THE MODEL**

In [164]:
model.fit(x_train,y_train)

LinearRegression()

WE USING LINER REGRESSTION AND THE FORMULA OF LINER REGRESSION IS
 Y= A+BX +E(WHERE "E" IS ERROR )
# THE LINE EQUATION IS Y= A+BX





In [165]:
model.intercept_,model.coef_

(np.float64(1.3045301408380539),
 array([-1.54751452e-01,  4.60117401e-02,  1.78842718e-01,  1.15737092e-05,
        -1.25384161e+00,  3.74746823e-01,  2.11069851e-01,  1.01482702e-01,
         1.91294429e+00, -1.53158595e+00,  4.14956838e-02,  1.42776633e+00,
         5.10123819e+00]))

**//PREDICTION**

In [166]:
y_pred = model.predict(x_test)

**NOW THE FINAL STEP IS EVALUATE**

In [167]:
from sklearn.metrics import mean_absolute_percentage_error
mean_absolute_percentage_error(y_test,y_pred)

0.05713709530845482

#  -> THE MAPE( mean absolute percentage error) for this problem using LinearRegression is 5.71%